File: jvm/leetcode_scala/src/main/scala/leetcode/editor/cn/[96]不同的二叉搜索树.scala
package leetcode.editor.cn

object UniqueBinarySearchTrees {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    def numTrees(n: Int): Int =
      impl().evalState(Context(1, n))

    private def impl(): State[Context, Int] =
      import StateOps.*
      get[Context].flatMap { s =>
        if s.low > s.high then point(1)
        else if s.memo.contains((s.low, s.high)) then point(s.memo((s.low, s.high)))
        else
          for
            v <- (s.low to s.high).foldLeft(point[Context, Int](0))((acc, num) =>
              for
                c     <- acc
                left  <- modify[Context](st => st.copy(low = s.low, high = num - 1)).flatMap(_ => impl())
                right <- modify[Context](st => st.copy(low = num + 1, high = s.high)).flatMap(_ => impl())
              yield c + left * right
            )
            _ <- modify[Context](st => st.copy(memo = st.memo.updated((s.low, s.high), v)))
          yield v
      }

    private case class Context(low: Int, high: Int, memo: Map[(Int, Int), Int] = Map.empty)
    import scala.util.control.TailCalls.*

    /** 栈安全的State Monad。可用于leetcode的用fp方式来解决问题。
      */
    private case class State[S, A](run: S => TailRec[(S, A)]) {
      def evalState(s: S): A = run(s).result._2

      def execState(s: S): S = run(s).result._1

      def map[B](ab: A => B): State[S, B] =
        flatMap(a => State(s => done((s, ab(a)))))

      def flatMap[B](afb: A => State[S, B]): State[S, B] =
        State(s =>
          tailcall(
            run(s).flatMap { case (s, a) => tailcall(afb(a).run(s)) }
          )
        )
    }

    private object StateOps {
      def point[S, A](a: A): State[S, A] = State(s => done((s, a)))

      def get[S]: State[S, S] = State(s => done((s, s)))

      def put[S](s: S): State[S, Unit] = State(_ => done((s, ())))

      def modify[S](ss: S => S): State[S, Unit] =
        for { s <- get[S]; _ <- put(ss(s)) } yield ()
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

@main def runUniqueBinarySearchTrees: Unit =
  println(
    UniqueBinarySearchTrees.Solution.numTrees(19)
  )

File: cpp/leetcode/editor/cn/[96]不同的二叉搜索树.cpp
#include <vector>
#include "util.h"
#include <map>
#include <stack>
#include <string>
#include <utility>
#include <memory>
#include <iterator>
#include <unordered_map>
#include <algorithm>
#include <numeric>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    int numTrees(int n) {
//        return (int) catalanNumber(n);
        for (int i = 0; i < n + 1; i++) {
            memo.emplace_back(n + 1, 0);
        }
        return count(1, n);
    }

    int64_t catalanNumber(int n) {
        auto numer = combinatoric(2*n,n);
        auto frag = n + 1;
        return numer / frag;
    }

    int64_t combinatoric(int64_t n, int64_t m) {
        if (n == m || m == 0) return 1;
        return combinatoric(n, m - 1) * (n - m + 1) / m;
    }

    int count(int low, int high) {
        if (low > high) return 1;

        if (memo[low][high] != 0) return memo[low][high];

        int res = 0;
        for (auto mid = low; mid <= high; mid++) {
            int left = count(low, mid - 1);
            int right = count(mid + 1, high);
            res += left * right;
        }
        memo[low][high] = res;
        return res;
    }

private:
    vector<vector<int>> memo;
};
//leetcode submit region end(Prohibit modification and deletion)
